torch/csrc/jit/tensorexpr/loopnest.h

```diff
@@ -206,6 +206,31 @@ class TORCH_API LoopNest {
   static void normalize(For* f, For** normalized);
   static bool flatten(const std::vector<For*>& f, For** flattened);
 
+  // Compresses the given buffer based on its use in the given Stmts.
+  // For example, given the input:
+  //
+  // for (int i = 0; i < 100; ++i) {
+  //   for (int j = 0; j < 200; ++j) {
+  //     A[i,j] = sin(i*j)
+  //   }
+  //   for (int j = 0; j < 199; ++j) {
+  //     B[i,j] = A[i,j] + A[i, j+1]
```

Comment:
Hm, I thought we also wanted to be able to compress to a non-1 size. E.g. in
```
for (int i = 0; i < 100; ++i) {
   for (int j = 0; j < 200; ++j) {
     A[i,j] = sin(i*j)
   }
   for (int j = 0; j < 199; ++j) {
     B[i,j] = A[i,j] + A[i, j+1] + A[i+1,j] + A[i+1, j+1]
    }
  }
```
We should be able to compress A from [100, 200] to [2, 200]. Similarly, if in the original example we peel+fuse loops, we'd like to be able to compress to [1, 2].

Reply:
Actually, I've tried to replay transformations mimicking `compute_at` at the innermost level and got stuck. How do we do this with fuse+compress? I'm looking at this IR as the input:
```
Original:
for (int i = 0; i < 100; ++i) {
  for (int j = 0; j < 200; ++j) {
    A[i,j] = sin(i*j)
  }
}
for (int i = 0; i < 100; ++i) {
  for (int j = 0; j < 199; ++j) {
    B[i,j] = A[i,j] + A[i, j+1]
  }
}
```
And I'd like to get to the following IR:
```
After compute_at(A,j2):
for (int i2 = 0; i2 < 100; ++i2) {
  for (int j2 = 0; j2 < 199; ++j2) {
    for (int t = 0; t < 2; t++) {
      A[i2, t] = sin(i1*(j2+t)) // dims: [100, 2]
    }
    B[i2,j2] = A[i2,0] + A[i2, 1]
  }
}
```

Reply:
This particular example has a loop-carried dependence on the `i` loop for buffer A. That is invalid in our IR AFAIK.

Reply:
Mm, yeah, but we still need to be able to represent it somehow since that's a canonical `compute_at` case.

Reply:
We'll definitely see this pattern in convolutions since we'll have something like:
```
compute pad:
 for n, h, w, c:
   pad[n, h, w, c] = ...
compute conv:
  for n, h, w, k:
    for c, r, s:
      conv[n, h, w, k] += pad[n, h+r, w+s, c] * filter[...]
```

Reply:
In my example we probably would also like to compress the `i2` dimension of `A` so that it becomes:
```
After compute_at(A,j2):
for (int i2 = 0; i2 < 100; ++i2) {
  for (int j2 = 0; j2 < 199; ++j2) {
    for (int t = 0; t < 2; t++) {
      A[0, t] = sin(i2*(j2+t)) // dims: [1, 2]
    }
    B[i2,j2] = A[0,0] + A[0, 1]
  }
}
```

Having absolutely no loop-carried dependencies seems like a very strong constraint, and I suspect we're already violating it - e.g. i'm pretty sure registerizer could also introduce a pattern like:
```
for i
   for j
      Allocate(t)
      // Use t
      Free(t)
```
This loop also has a cross-iteration dependencies, e.g. it cannot be parallelized across `i`, but we are probably still fine generating a correct code for it. We might need to think about how to prevent applying transformations that might need to know about that though (e.g. "parallel").

Reply:
If cross iteration dependences are necessary we just have to make sure that we carefully handle all transformations with that in mind. 

For this particular example, with fusion and buffer compression we can get to:
```
for (int i = 0; i < 100; ++i) {
  for (int j = 0; j < 200; ++j) {
    A[0,j] = sin(i*j)
  }
  for (int j = 0; j < 199; ++j) {
    B[i,j] = A[0,j] + A[0, j+1]
  }
}
```
We need something more to get to your desired form.

Reply:
Since that is orthogonal to this PR, added a new issue to track that. We can discuss it there. 
https://github.com/pytorch/pytorch/issues/55933

Reply:
To answer your first question here:

> I thought we also wanted to be able to compress to a non-1 size

Yes, that needs to be done. I will do that as a follow up PR.